### PR TITLE
#2136 performance improvement 2 step mappings

### PR DESCRIPTION
**Pull request by sjaakd**

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java [hunk 1]
```diff
@@ -376,32 +377,56 @@ private Assignment resolveViaMethod(Type sourceType, Type targetType, boolean co
...
+         *
+         * @param sourceType the source type
+         * @param targetType the target type
+         * @param method     the method to match
+         * @return an assignment if a match, given the criteria could be found. When the method is a
+         * buildIn method, all the bookkeeping is applied.
+         */
+        private Assignment applyMatching(Type sourceType, Type targetType, Method method) {
+
+            if ( method instanceof SourceMethod ) {
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java:391]: What if the method is not one of these types? It can be an intermediate `ForgedMethod` or an existing `HelperMethod` that we have created. Why shouldn't we reuse that one?

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java [hunk 2]
```diff
@@ -486,10 +511,9 @@ private Assignment resolveViaConversionAndMethod(Type sourceType, Type targetTyp
                 ySourceType = ySourceType.resolveTypeVarToType( targetType, methodYCandidate.getResultType() );
 
                 if ( ySourceType != null ) {
-                    methodRefY = resolveViaMethod( ySourceType, targetType, true );
+                    methodRefY = applyMatching( ySourceType, targetType, methodYCandidate );
                     if ( methodRefY != null ) {
-                        Type targetTypeX = methodYCandidate.getSourceParameters().get( 0 ).getType();
-                        ConversionAssignment conversionXRef = resolveViaConversion( sourceType, targetTypeX );
+                        ConversionAssignment conversionXRef = resolveViaConversion( sourceType, ySourceType );
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java:516]: Why is this change needed?

`ySourceType` and `targetTypeX` don't have to be the same types I think. One can be more specific then the other.

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java [hunk 3]
```diff
@@ -376,32 +377,56 @@ private Assignment resolveViaMethod(Type sourceType, Type targetType, boolean co
...
+        /**
+         * Applies matching to given method only
+         *
+         * @param sourceType the source type
+         * @param targetType the target type
+         * @param method     the method to match
+         * @return an assignment if a match, given the criteria could be found. When the method is a
+         * buildIn method, all the bookkeeping is applied.
+         */
+        private Assignment applyMatching(Type sourceType, Type targetType, Method method) {
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java:389]: Perhaps this should be named `resolveViaGivenMethod`? Or something in that direction. `applyMatching` differs from the other methods in my opinion

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java (hunk 1 above)
Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java:391]: is not possible here.. We don't resolve forged methods.. If so, we should throw an unsupported.

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java (hunk 2 above)
Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java:516]: the change was forgotten last time.. it's actually a bug.

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java (hunk 3 above)
Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java:389]: not entirely.. We still have to apply all criteria to really figure out if it matches (qualifiers, xml-stuff, resultTypes, etc)

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java (hunk 1 above)
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java:391]: You are right

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java (hunk 2 above)
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java:516]: 👍

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java (hunk 3 above)
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java:389]: OK

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java (hunk 1 above)
Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java:391]: resolved, I guess?
